src/main/java/com/zufar/icedlatte/auth/google/endpoint/AuthEndpoint.java

```diff
@@ -0,0 +1,189 @@
+            + "/callback";
+    TokenResponse token = flow.newTokenRequest(code).setRedirectUri(redirectAuthUri).execute();
+
+    HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+    JsonFactory jsonFactory = GsonFactory.getDefaultInstance();
+    GoogleIdTokenVerifier verifier =
+        new GoogleIdTokenVerifier.Builder(httpTransport, jsonFactory)
+            .setAudience(Collections.singletonList(clientId))
+            .build();
+
+    // verify idToken
+    GoogleIdToken idToken = verifier.verify((String) token.get("id_token"));
+    if (idToken != null) {
+      GoogleIdToken.Payload payload = idToken.getPayload();
+      String email = payload.getEmail();
+      log.info("User email: {}", email);
+
+      // find or create user
+      String password = findOrCreateUser(email);
+      log.info("Success logged with email={}", email);
+
+      // login in main form
+      sendPostRequest(authenticateUri, new UserAuthenticationRequest(email, password));
+
+      // redirect on main page
+      return ResponseEntity.status(HttpStatus.TEMPORARY_REDIRECT)
+          .header("Location", redirectUri)
+          .build();
+    } else {
+      log.error("Invalid ID token.");
+      return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
+    }
+  }
+
+  private GoogleAuthorizationCodeFlow initializeFlow()
+      throws IOException, GeneralSecurityException {
+    HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+    JsonFactory jsonFactory = GsonFactory.getDefaultInstance();
+    return new GoogleAuthorizationCodeFlow.Builder(
+            httpTransport, jsonFactory, clientId, clientSecret, List.of(scope))
+        .setAccessType("offline")
+        .setApprovalPrompt("force")
+        .build();
+  }
+
+  private String findOrCreateUser(String email) {
+    Optional<UserEntity> userEntity = userCrudRepository.findByEmail(email);
+    if (userEntity.isEmpty()) {
+      String password = UUID.randomUUID().toString();
+      userRegistrationService.register(new UserRegistrationRequest(email, email, email, password));
+      return password;
+    } else {
+      return userEntity.get().getPassword();
+    }
+  }
+
+  public static void sendPostRequest(
+      String authenticateUri, UserAuthenticationRequest userAuthenticationRequest) throws InterruptedException {
+    HttpClient client = HttpClient.newHttpClient();
+
+    Gson gson = new Gson();
+    String json = gson.toJson(userAuthenticationRequest);
+
+    HttpRequest request =
+        HttpRequest.newBuilder()
+            .uri(URI.create(authenticateUri))
+            .header("Content-Type", "application/json")
+            .POST(HttpRequest.BodyPublishers.ofString(json))
+            .build();
+    try {
+      client.send(request, HttpResponse.BodyHandlers.ofString());
+    } catch (IOException | InterruptedException e) {
+      log.error("Error sending request", e);
+      throw new InterruptedException();
+    }
+  }
```

Comment:
nit: how about considering [`WebClient`](https://docs.spring.io/spring-framework/reference/web/webflux-webclient.html)? No push, I'm just wondering if it could bring us any benefits :) Looks like _sendPostRequest_ method's body would be like:

```java
@Autowired
private WebClient webClient;
    ...
return webClient.post()
            .uri(authenticateUri)
            .header("Content-Type", "application/json")
            .body(BodyInserters.fromValue(userAuthenticationRequest))
            .retrieve() // Start the retrieval of the response body
            .bodyToMono(String.class) // Convert the response body to String
            .doOnError(e -> log.error("Error sending request", e));
```

Reply:
for me its equal